---
title: "Protein Classification"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Protein Classification}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction
The motivation behind this package was to create a way to visualize and interpret various protein structures. The current way to visualize the makeup of a protein sequence was difficult as it is time intensive to clean up the original file. Typically in a PDB file, there are various water molecules, nucleic acids and bound molecules such as sugars or inhibitors. Using `protein5555`, a simple way has been implemented to input a PDB file and return the longest protein string to visualize and analyze. The function can quickly identify the protein sequence and filter out the unnecessary data. 

Protein strings are made up of 20 different amino acids with varying length. A function has been written to help classify each of these amino acids into meaningful categories such as polarity and the type of charge. Polarity is an important analytical tool, as it determines if the amino acid is hydrophobic or hydrophilic, which then determines the protein size and type of bonds. The type of charge, either positive or negative, can determine protein stability or the protein bonds. 

Various plotting functions were implemented in this package to visualize the amount of amino acids in each category from the classify function. There is an option to compare between protein families, such as kinase and polymerase and view the difference in each of these protein sequences. This package offers a great introduction into the protein sequence makeup. 

# Installation
The `protein5555` package has package dependencies on `bio3d`, `RColorBrewer`, 
`dplyr` and `ggplot2`. These should be installed if not downloaded prior. To use the `protein5555` package, you'll need to install it from GitHub using the `devtools` package. This can be done by following this link to the GitHub repository: https://github.com/EmelineHaroldsen/protein5555.git

Once installed, you can load the protein5555 package in your R session using the following command:

```{r}
library(protein5555)
```

# Usage with Example Data

In this section, we'll demonstrate how to use the `protein5555` package to analyze protein structures and perform comparative analyses. We will show this using the built in datasets in this package, however using your own PDB file is compatible with this package. 

## Description of Datasets

The `protein5555` package has 6 built-in datasets (`bqr`, `fpu`, `huo`, `qgy`, `r7g`, and `sl0`). Each dataset contains a protein sequence. We used PolBase to identify various polymerase protein sequences and then viewed and downloaded the respective PDB files on RSCB PDB. Polymerases are a type of protein that synthesize nucleic acid (for example: DNA Pol 1 in bacteria is responsible for nick translation and DNA Pol 3 in bacteria does the bulk of the work necessary for replcating the bacterial genome). Furthermore, polymerases often have some kind of proofreading activity. The three polymerase datasets are `bqr`, `huo` and `sl0`. 

The other three sequences, `fpu`, `qgy` and `r7g` came from the KLIFS database. KLIFS was a resource to identify kinase protein structures and were also downloaded from RSCB PDB. Kinases are a type of protein that adds phosphoryl groups to Thr, Tyr, or Ser residues on other proteins (or sometimes itself). These modifications have major implications in cellular signalling pathways. (For example, the rhodoposin kinase modifies the protein rhodoposin, after it's retinol has absorbed light. Rhodoposin is a key protein for sight. Phosophorylated rhodoposin is inactive, and no longer able to transmit signals further down the signalling cascade. The activity of rhodoposin kinase is important for preventing sudden intense beams of light (such as from the headlights of a car driving by) from blinding us.)

Links to the database sources are included below:

RSCB PDB: https://www.rcsb.org/

PolBase: https://polbase.neb.com/

KLIFS: https://klifs.net/

## Load Example Datasets

Below, we have loaded the RData files from this package and called a constructor to assign the `protein5555` class to each protein. Furthermore, called a constructor to assign the `protein5555_list` class to a list of each protein object. 
```{r}
a <- protein5555(bqr, "pol")
b <- protein5555(fpu, "kin")
d <- protein5555(huo, "pol")
e <- protein5555(qgy, "kin")
f <- protein5555(r7g, "kin")
g <- protein5555(sl0, "pol")

kin_pol <- protein5555_list(list(a, b, d, e, f, g))
```

## Summary Statistics
There is a built in summary statistic of the function that is compatible with the `protein5555_list` class.  
If the summary is of an individual protein the summary function returns the family of the protein and some statistics about the proportion of amino acids of the protein sequence which are polar, nonpolar, charged, etc.
If the summary is of protein list, it will return the families in the list, and some summary statistics about the proportion of amino acids (in the whole list) which are polar, nonpolar, charged, etc.
```{r}
summary(a)
summary(kin_pol)
```

## Plot Amino Acid Proportions
There are several plotting functions. Below, we show plotting the respective protein families, (polymerase and kinase) and the amino acid frequencies in each family. As an example, we also plot the amino acid frequencies of a single protein.

Comparing the amino acid frequencies between the kinase and polymerase families allows us to generate hypotheses about what properties are important for each families respective protein functioning. For example, in the below plot it is seen that kinases have more Tyr, Thr, and Ser residues then polymerases, which is interesting, because these are the residues that kinases modify. This allows us to hypothesis that kinase regulation may often involve phosphorlyation, particularly self-phosphorylation, while polymerase activity may be not regulated as often by kinase activity.
It's also worthwhile interesting to note from both the family plot and the individual protein plot, kinases and polymerases do not have many cysteines (although kinases are appear to have more cysteines then polymerases). This isn't too shocking, as cysteines are highly susceptible to oxidative damage and are often rare in protein sequences.
```{r, warning=FALSE, fig.width=8, fig.height=6}
plot(kin_pol)
plot(a)
```

## Plot by Charge or Polarity
The charge and polarity plots that allow you to compare the percentage of amino acids within protein families that are charged or polar.

For example, below we compare the frequency of amino acids that are charged in kinases and polymerases, as well as the frequency of amino acids that are polar/nonpolar.

The polymerase dataset has more positively charged residues than kinase dataset - which is explained because polymerases must be able to bind to a negatively charged nucleic acid backbone. However, the polymerase dataset also have more NEGATIVELY charged residues then the kinase dataset. At first, this seems surprising, since negatively charged residues (depending on their placement), would repulse the nucleic acid backbone. However, polymerases must coordinate magnesium, which is a positively charged metal ion, in order to be active. The higher frequency of negative residues in polymerases may reflect the residues that polymerases use in order for magnesium coordination.

Interestingly, although polymerases tend to have more charged residues then kinases, kinases and polymerases have very similar frequencies of polar and nonpolar residues. This may indicate more about properties that required for proper protein folding than about either family specifically. In other words, proper protein folding requires nonpolar and polar residues. It's possible that proper protein folding requires a particular proporition of nonpolar and polar residues, but assessing this would require looking at many more protein families.
```{r, fig.width=8, fig.height=6}
plot(kin_pol, kind = "charge")

plot(kin_pol, kind = "polar")
```


## Compare Sequence Lengths Between DNA Polymerases and Kinases
With the kind argument set to "length", we can visualize a boxplot of the total sequence lengths of each protein family.

Although as a whole, size is not terribly informative protein feature, there may still be some information which can be glean from plotting and comparing protein family sizes.

In particular, larger proteins may be easier to regulate via modification, or other protein binding (since if the protein is larger, it will have more space for proteins to dock on it). In the below example plot, we see that kinases have a greater variance in sequence length then polymerases, and the median of polymerase sequence length is greater than kinase length. This could be a result of the above the hypothesis, indicating that polymerases are generally more highly regulated then kinases.
```{r, fig.width=8, fig.height=6}
plot(kin_pol, kind = "length")
```

# Using Data to Clean and Classify
There are functions built into the `protein5555` package that allow for a PDB file to be entered (or a variable assigned to a previously read in PDB file) and the longest protein string to be selected. There are typically two strings, labeled A or B, and the `cleaner` function will return which string was selected while storing the longer protein sequence. The classify sequence allows for the various 20 amino acids to be classified with polarity and the type of charge. 

The `cleaner` function returns a character vector of the longest string. For the example below, `huo1` is used as the input, which is a previously read in PDB file under this variable name and contains raw data. The example below shows that string A was the longer string and returns the individual amino acids from that string. 

The `classify` function will return a data frame, so the head function will be implemented to show a small example. This should return the individual amino acids with their respective labels such as "polar", "nonpolar", "positive" or "negative".
```{r}
head(cleaner(huo1))
head(classify(huo))
```

# Conclusion
In conclusion, the `protein5555` package benefits individuals seeking to navigate the intricacies of protein structure data sourced from PolBase and KLIFS databases in the PDB file format. 

The utility of this package stems from its adeptness in classifying proteins into distinct families, enabling comprehensive comparisons of their structural attributes. Furthermore, it offers robust tools for scrutinizing amino acid proportions, thereby augmenting the depth of analysis. The package plotting functions tailored to illuminate the polarity and charge distribution of individual amino acids are great visual aids in investigating protein composition and behavior.

